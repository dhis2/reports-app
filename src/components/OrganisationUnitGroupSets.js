/* React */
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

/* d2-ui components */
import { DropDown } from '@dhis2/d2-ui-core'

/* App context */
import AppContext from '../pages/AppContext'

/* i18n */
import i18n from '@dhis2/d2-i18n'

/* styles */
import styles from '../utils/styles'

export class OrganisationUnitGroupSets extends PureComponent {
    static propTypes = {
        d2: PropTypes.object.isRequired,
        onChange: PropTypes.func.isRequired,
        values: PropTypes.object.isRequired,
        dropdownStyle: PropTypes.object,
        fullWidth: PropTypes.bool,
    }

    static defaultProps = {
        fullWidth: true,
        dropdownStyle: {
            display: 'block',
        },
    }

    constructor() {
        super()

        this.state = {
            organisationUnitGroupSets: [],
        }
    }

    componentDidMount() {
        const d2 = this.props.d2
        d2.models.organisationUnitGroupSet
            .list({
                paging: false,
                fields: 'id,displayName,organisationUnitGroups[id,displayName]',
            })
            .then(organisationUnitGroupSetsResponse => {
                this.setState({
                    organisationUnitGroupSets: organisationUnitGroupSetsResponse.toArray(),
                })
            })
            .catch(() => {
                // TODO Manage error
            })
    }

    handleOrganisationUnitGroupSetChange = organisationUnitGroupSetId => element => {
        this.props.onChange(organisationUnitGroupSetId, element)
    }

    renderOrganisationUnitGroupSetDropdown = organisationUnitGroupSet => (
        <div key={organisationUnitGroupSet.id}>
            <span style={styles.formLabel}>
                {organisationUnitGroupSet.displayName}
            </span>
            <DropDown
                style={this.props.dropdownStyle}
                fullWidth={this.props.fullWidth}
                value={this.props.values[organisationUnitGroupSet.id]}
                onChange={this.handleOrganisationUnitGroupSetChange(
                    organisationUnitGroupSet.id
                )}
                menuItems={organisationUnitGroupSet.organisationUnitGroups}
                includeEmpty
                emptyLabel={i18n.t('Select Option')}
                hintText={i18n.t('Select Option')}
            />
        </div>
    )

    render = () =>
        this.state.organisationUnitGroupSets.map(organisationUnitGroupSet =>
            this.renderOrganisationUnitGroupSetDropdown(
                organisationUnitGroupSet
            )
        )
}

export default props => (
    <AppContext.Consumer>
        {appContext => (
            <OrganisationUnitGroupSets d2={appContext.d2} {...props} />
        )}
    </AppContext.Consumer>
)
