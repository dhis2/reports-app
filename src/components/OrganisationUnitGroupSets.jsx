/* React */
/* i18n */
import i18n from '@dhis2/d2-i18n'
/* d2-ui components */
import { DropDown } from '@dhis2/d2-ui-core'
import { CircularProgress } from '@material-ui/core'
import PropTypes from 'prop-types'
import React from 'react'
/* App context */
import { connect } from 'react-redux'
import { selectOrgUnitOption } from '../redux/actions/organisationUnits.js'
import { formLabel } from '../utils/styles/shared.jsx'
import styles from './OrganisationUnitGroupSets.module.css'

const createGroupSetOnChange = (groupSetId, onChange) => (event) => {
    onChange(groupSetId, event.target.value)
}

const labelText = i18n.t('Select Option')

const OrganisationUnitGroupSetDropdown = ({
    groupSet,
    fullWidth,
    value,
    onChange,
}) => (
    <div key={groupSet.id}>
        <span className={formLabel.className}>{groupSet.displayName}</span>
        <DropDown
            fullWidth={fullWidth}
            value={value}
            onChange={onChange}
            menuItems={groupSet.organisationUnitGroups}
            includeEmpty
            emptyLabel={labelText}
            hintText={labelText}
        />
        {formLabel.styles}
    </div>
)

OrganisationUnitGroupSetDropdown.propTypes = {
    fullWidth: PropTypes.bool.isRequired,
    groupSet: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string,
}

OrganisationUnitGroupSetDropdown.defaultProps = {
    value: '',
}

export const __OrganisationUnitGroupSets = ({
    groupSets,
    isLoading,
    selectGroupSet,
    values,
    fullWidth,
}) =>
    isLoading ? (
        <div className={styles.loader}>
            <CircularProgress size={24} thickness={3} />
        </div>
    ) : (
        groupSets.map((groupSet) => (
            <OrganisationUnitGroupSetDropdown
                groupSet={groupSet}
                key={groupSet.id}
                onChange={createGroupSetOnChange(groupSet.id, selectGroupSet)}
                value={values[groupSet.id]}
                fullWidth={fullWidth}
            />
        ))
    )

__OrganisationUnitGroupSets.propTypes = {
    groupSets: PropTypes.array.isRequired,
    isLoading: PropTypes.bool.isRequired,
    selectGroupSet: PropTypes.func.isRequired,
    values: PropTypes.object.isRequired,
    fullWidth: PropTypes.bool,
}

__OrganisationUnitGroupSets.defaultProps = {
    fullWidth: true,
}

const mapStateToProps = (state) => ({
    isLoading: state.orgUnitGroupSets.loading,
    groupSets: state.orgUnitGroupSets.collection,
    values: state.organisationUnits.selectedOptions,
})

export const OrganisationUnitGroupSets = connect(mapStateToProps, {
    selectGroupSet: selectOrgUnitOption,
})(__OrganisationUnitGroupSets)
