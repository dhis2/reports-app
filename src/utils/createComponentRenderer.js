import React from 'react'
import { shallow } from 'enzyme'

export default function createComponentRenderer(Component, defaultProps) {
    return function(customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
