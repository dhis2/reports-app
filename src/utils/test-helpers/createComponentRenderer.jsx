import { shallow } from 'enzyme'
import React from 'react'

export default function createComponentRenderer(Component, defaultProps) {
    return function (customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
