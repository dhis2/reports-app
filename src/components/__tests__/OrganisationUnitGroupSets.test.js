import createComponentRenderer from '../../utils/test-helpers/createComponentRenderer'
import { OrganisationUnitGroupSets } from '../OrganisationUnitGroupSets'

jest.mock('@dhis2/d2-ui-core', () => ({
    DropDown: 'DropDown',
}))

describe('Test <OrganisationUnitGroupSets /> rendering:', () => {
    const groupSets = [
        {
            id: 'organisationUnitGroupSet1',
            displayName: 'organisationUnitGroupSet1',
            organisationUnitGroups: [
                {
                    id: 'item1',
                    displayName: 'item1',
                },
            ],
        },
        {
            id: 'organisationUnitGroupSet2',
            displayName: 'organisationUnitGroupSet2',
            organisationUnitGroups: [
                {
                    id: 'item1',
                    displayName: 'item1',
                },
            ],
        },
    ]
    const selectGroupSet = jest.fn()
    const defaultProps = {
        groupSets,
        isLoading: false,
        selectGroupSet,
        values: {},
    }

    const componentRenderer = createComponentRenderer(
        OrganisationUnitGroupSets,
        defaultProps
    )

    it('It should match the snapshot', () => {
        expect(componentRenderer()).toMatchSnapshot()
    })

    it('Should render a loader when not ready', () => {
        expect(componentRenderer({ isLoading: true })).toMatchSnapshot()
    })
})
