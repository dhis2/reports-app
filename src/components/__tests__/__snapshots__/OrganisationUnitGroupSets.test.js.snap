// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <OrganisationUnitGroupSets /> rendering: It should match the snapshot 1`] = `
Array [
  <OrganisationUnitGroupSetDropdown
    fullWidth={true}
    groupSet={
        Object {
            "displayName": "organisationUnitGroupSet1",
            "id": "organisationUnitGroupSet1",
            "organisationUnitGroups": Array [
              Object {
                "displayName": "item1",
                "id": "item1",
              },
            ],
          }
    }
    key="organisationUnitGroupSet1"
    onChange={[Function]}
    value=""
/>,
  <OrganisationUnitGroupSetDropdown
    fullWidth={true}
    groupSet={
        Object {
            "displayName": "organisationUnitGroupSet2",
            "id": "organisationUnitGroupSet2",
            "organisationUnitGroups": Array [
              Object {
                "displayName": "item1",
                "id": "item1",
              },
            ],
          }
    }
    key="organisationUnitGroupSet2"
    onChange={[Function]}
    value=""
/>,
]
`;

exports[`Test <OrganisationUnitGroupSets /> rendering: Should render a loader when not ready 1`] = `
<div>
  <WithStyles(CircularProgress)
    size={24}
    thickness={3}
  />
  <style>
    
                    div {
                        padding-top: 16px;
                        text-align: center;
                    }
                
  </style>
</div>
`;
